src/ampform/kinematics.py

```diff
@@ -271,84 +298,421 @@ def __recursive_helicity_angles(  # pylint: disable=too-many-locals
                     )
                     helicity_angles.update(angles)
 
-        return DataSet(helicity_angles)
+        return helicity_angles
 
     initial_state_id = next(iter(topology.incoming_edge_ids))
     initial_state_edge = topology.edges[initial_state_id]
     assert initial_state_edge.ending_node_id is not None
     return __recursive_helicity_angles(
-        events, initial_state_edge.ending_node_id
+        four_momenta, initial_state_edge.ending_node_id
     )
 
 
-def _get_boost_z_matrix(beta: ScalarSequence) -> MatrixSequence:
-    n_events = len(beta)
-    gamma = 1 / np.sqrt(1 - beta ** 2)
-    zeros = np.zeros(n_events)
-    ones = np.ones(n_events)
-    return MatrixSequence(
-        np.array(
+def compute_invariant_masses(
+    four_momenta: FourMomentumSymbols, topology: Topology
+) -> Dict[str, sp.Expr]:
+    """Compute the invariant masses for all final state combinations."""
+    if topology.outgoing_edge_ids != set(four_momenta):
+        raise ValueError(
+            f"Momentum IDs {set(four_momenta)} do not match "
+            f"final state edge IDs {set(topology.outgoing_edge_ids)}"
+        )
+    invariant_masses = {}
+    for state_id in topology.edges:
+        attached_state_ids = determine_attached_final_state(topology, state_id)
+        total_momentum = ArraySum(
+            *[four_momenta[i] for i in attached_state_ids]
+        )
+        invariant_mass = InvariantMass(total_momentum)
+        name = get_invariant_mass_label(topology, state_id)
+        invariant_masses[name] = invariant_mass
+    return invariant_masses
+
+
+class ArraySum(sp.Expr):
```

Comment:
If I understand correctly this would be some code that will end up in sympy eventually right? I would put all of that code into a separate file to make this a bit easier to clean up later on as well

Reply:
I have to think a bit about this one. It's unlikely that it will end up in SymPy, because it should theoretically be possible to do this with existing classes https://github.com/sympy/sympy/discussions/22269. So I put the classes that most likely won't end up in SymPy under kinematics.